metronome/hotstuff/consensus/src/metronome/hotstuff/consensus/basic/ProtocolState.scala

```diff
@@ -0,0 +1,492 @@
+  }
+
+  private def handleQuorum(
+      event: Validated[MessageReceived[A]],
+      phase: VotingPhase
+  )(
+      f: Quorum[A] => Transition[A]
+  ): PartialFunction[Message[A], TransitionAttempt[A]] = {
+    case m: Quorum[_]
+        if matchingLeader(event) &&
+          m.quorumCertificate.viewNumber == viewNumber &&
+          m.quorumCertificate.phase == phase &&
+          m.quorumCertificate.blockHash == preparedBlockHash =>
+      Right(f(m))
+
+    case m: Quorum[_]
+        if matchingLeader(event) &&
+          m.quorumCertificate.viewNumber == viewNumber &&
+          m.quorumCertificate.phase == phase &&
+          m.quorumCertificate.blockHash != preparedBlockHash =>
+      Left(UnexpectedBlockHash(event, preparedBlockHash))
+  }
+
+  /** Categorize unexpected messages into ones that can be re-queued or discarded.
+    *
+    * At this point we already know that the messages have been validated once,
+    * so at at least they are consistent with their own view, e.g. sending to the
+    * leader of their own view.
+    */
+  private def handleUnexpected(e: MessageReceived[A]): ProtocolError[A] = {
+    e.message match {
+      case m: NewView[_] if m.viewNumber >= viewNumber =>
+        TooEarly(e, m.viewNumber.next, Phase.Prepare)
+
+      case m: Prepare[_] if m.viewNumber > viewNumber =>
+        TooEarly(e, m.viewNumber, Phase.Prepare)
+
+      case m: Vote[_]
+          if m.viewNumber > viewNumber ||
+            m.viewNumber == viewNumber && m.phase.isAfter(phase.prev) =>
+        TooEarly(e, m.viewNumber, m.phase.next)
+
+      case m: Quorum[_]
+          if m.quorumCertificate.viewNumber > viewNumber ||
+            m.quorumCertificate.viewNumber == viewNumber &&
+            m.quorumCertificate.phase.isAfter(phase.prev) =>
+        TooEarly(e, m.viewNumber, m.quorumCertificate.phase.next)
+
+      case _ =>
+        Unexpected(e)
+    }
+  }
+
+  /** Try to match a message to expectations, or return Unexpected. */
+  private def matchingMsg(e: MessageReceived[A])(
+      pf: PartialFunction[Message[A], TransitionAttempt[A]]
+  ): TransitionAttempt[A] =
+    pf.lift(e.message).getOrElse(Left(handleUnexpected(e)))
+
+  /** Check that a message is coming from the view leader and is for the current phase. */
+  private def matchingLeader(e: MessageReceived[A]): Boolean =
+    e.message.viewNumber == viewNumber &&
+      e.sender == federation.leaderOf(viewNumber)
+
+  /** Broadcast a message from the leader to all replicas.
+    *
+    * This includes the leader sending a message to itself,
+    * because the leader is a replica as well. The effect
+    * system should take care that these messages don't
+    * try to go over the network.
+    *
+    * NOTE: Some messages trigger transitions; it's best
+    * if the message sent to the leader by itself is handled
+    * before the other messages are sent out to avoid any
+    * votes coming in return coming in phases that don't
+    * yet expect them.
+    */
```

Comment:
Exactly! I made this comment before I added `TooEarly` to help with this buffering, but then I thought it probably makes sense anyway to carry out local effects first, then we're double protected. It can't handle the case of `NewView` arriving from a replica that timed out, so `TooEarly` seems necessary, but if it's easy to do there wouldn't be any harm in it, it would just act on things we already know, and most likely it wouldn't delay sending out messages by much. 

We'll see, if it does, we can always just rely on buffering, but at least the comment can remind us of the option.